velox/functions/prestosql/MapFromEntries.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/expression/VectorReaders.h"
+#include "velox/expression/VectorWriters.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/map.html
+class MapFromEntriesFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 1);
+    auto& arg = args[0];
+    MapVectorPtr mapVector;
+
+    auto inputArray = arg->as<ArrayVector>();
+    auto rowVector = inputArray->elements();
+    auto rowKeys = rowVector->as<RowVector>()->childAt(0);
```

Comment:
The elements vector may have encodings. It doesn't have to be flat, hence, the function needs to handle all encodings.

Reply:
Do you mean we also want to handle both flat and constant encoding for the elements vector (RowVector)? I thought handling the top level (single-arg of array vector) should suffice.

Reply:
The top-level argument is guaranteed to be either flat or constant, but nested vectors can have any encoding (flat, constant or dictionary).

Reply:
I may need a quick chat to help understand how to handle such nested vector encodings in this case.

Since this udf takes one argument: `array(row(K, V))`, the encodings will be something like below, correct? 

| Level  | variable | vector encoding|
| ------------- | ------------- | ------------- |
| top level  | args[0] | constant / flat < ComplexType >  |
| second level  | *topArrayVectorPtr->elements() | constant / flat / dict < ComplexType >  |

Why aren't the vector encodings just `array` and `row`? Appreciate if there is any document to help me understand the mappings between vector types and encodings.

I can find the [example code](https://github.com/darrenfu/velox/blob/dd955aaddff0601a23c56d1ad62d0917b3aa2f94/velox/functions/prestosql/MapKeysAndValues.cpp#L35-L47) for the top level constant/flat handling. But for the second level (e.g. row vector), there are no similar examples. Is it something like below? How shall we handle dict vector encoding? 

```cpp
auto inputArray = decodedArg->base()->as<ArrayVector>();
auto rowVector = inputArray->elements();
switch (rowVector->encoding()) {
  case CONSTANT: {
    // repeat the same logic where top level vector is a constant vector
    auto* constantRow = arg->as<ConstantVector<ComplexType>>();
    const auto& flatRow = constantRow->valueVector();
    ...
    break;
  }
  case FLAT: {
    auto actualRowVector = rowVector->as<RowVector>();
    ...
    break;
  }
  case DICTIONARY:
    // How to handle this encoding? 
    break;
}
```

Reply:
@darrenfu Darren, feel free to ping me offline or book some VC time.

Encodings and types are independent: https://facebookincubator.github.io/velox/develop/vectors.html 

A canonical way to handle encodings is to use DecodedVector. In Vector functions, we often introduce fast paths for common use cases.

Reply:
Thanks @mbasmanova ! I booked 30 min on your calendar. Let's sync to clear some of my roadblocks to accelerate velox development then!